Add a new command to get the issues filtered by their state

NotBad4U opened the pull request.

NotBad4U committed:
Add new command to get the issues filtered by their state: open, close, reopen

divarvel commented on src/main.rs:
```diff
@@ -32,11 +32,28 @@ use std::error::Error;
+        match s.to_lowercase().as_ref() {
+            "open" => Ok(MyIssueState(IssueState::Opened)),
+            "close" => Ok(MyIssueState(IssueState::Closed)),
+            "reopen" => Ok(MyIssueState(IssueState::Reopened)),
```
can you rename it to `reopened` to be consistent with the enum?

divarvel commented on src/main.rs:
```diff
@@ -32,11 +32,28 @@ use std::error::Error;
+            "open" => Ok(MyIssueState(IssueState::Opened)),
+            "close" => Ok(MyIssueState(IssueState::Closed)),
+            "reopen" => Ok(MyIssueState(IssueState::Reopened)),
+            _ => Err(format!("Unknow state: {}", s)),
```
s/Unknow/Unknown/

divarvel commented on src/main.rs:
```diff
@@ -181,6 +198,24 @@ fn get_user_id_by_name(name: &str) -> Result<UserId, Box<Error>> {
+        .and_then(|issues| {
+            issues.into_iter()
+            .filter(|i| i.state == filter_state.0)
+            .for_each(|i| println!("#{} {:?} {} {}", i.id, i.state, i.title, i.created_at.format("%F %H:%M")));
```
Not sure about how it's usually done in Rust, but relying on `Debug` for regular display seems a bit odd.
Poke @Geal

NotBad4U replied:
I can implement the `Trait Display` for `MyIssueState` to correct this if you want ?

Geal replied:
yes, `Debug` is not meant for the user side

divarvel commented on src/main.rs:
```diff
@@ -287,6 +329,12 @@ enum Cmd {
+    #[structopt(name = "l", about = "List all gitlab issues")]
+    ListIssues {
+        #[structopt(name = "filter", short = "f", long = "filter",
+                    help = "Filter the issues by state. Possible values are: open, close, reopen")]
```
see first comment about "reopened"

divarvel reviewed (changes requested):
Hi, apart from a few typos here and there and the use of Debug for regular display, LGTM

Commit:
Impl display trait for MyIssueState and fix typos

divarvel commented on src/main.rs:
```diff
@@ -32,11 +32,49 @@ use std::error::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_ref() {
+            "open" => Ok(MyIssueState(IssueState::Opened)),
+            "close" => Ok(MyIssueState(IssueState::Closed)),
```
s/close/closed/

divarvel commented on src/main.rs:
```diff
@@ -32,11 +32,49 @@ use std::error::Error;
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self.0 {
+            IssueState::Opened => write!(f, "open"),
+            IssueState::Closed => write!(f, "close"),
```
s/close/closed/

divarvel commented on src/main.rs:
```diff
@@ -287,6 +350,12 @@ enum Cmd {
+    #[structopt(name = "l", about = "List all gitlab issues")]
+    ListIssues {
+        #[structopt(name = "filter", short = "f", long = "filter",
+                    help = "Filter the issues by state. Possible values are: open, close, reopened")]
```
s/close/closed/

divarvel committed:
Fix typos in `IssueState` text representation

divarvel committed:
Make "open" default value for `--filter` in `l` cmd

This is a sensible default, and the main goal of this tool is brevity.

divarvel committed:
Extract url generation

divarvel committed:
Display `iid` instead of `id` in issues list

`id` is an internal id, unique across projects, not exposed in the UI.
`iid` is what we want: it's the id displayed in the UI and visible in
URLs.

divarvel commented:
I'm fixing a few issues and adding some bells and whistles.

divarvel committed:
Display issue url in issue list

divarvel commented:
@NotBad4U I've fixed the typos, added a default value for `--filter`, displayed the URLs in the list output. Looks good to you?

NotBad4U commented:
Look good to me too 👍
